keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/TreeBuilder.java

```diff
@@ -0,0 +1,433 @@
+            STARTING_DELTA_LIKELIHOOD,
+            STARTING_TREE_SIZE
+        );
+    }
+
+    public static TreeBuilder buildOtherHalfOfTree(TreeBuilder currentTree,
+                                                   List<Vertex<DoubleTensor>> latentVertices,
+                                                   List<Vertex> probabilisticVertices,
+                                                   LogProbGradientCalculator logProbGradientCalculator,
+                                                   final List<? extends Vertex> sampleFromVertices,
+                                                   double logU,
+                                                   int buildDirection,
+                                                   int treeHeight,
+                                                   double epsilon,
+                                                   double logOfMasterPMinusMomentumBeforeLeapfrog,
+                                                   KeanuRandom random) {
+
+        TreeBuilder otherHalfTree;
+
+        if (buildDirection == -1) {
+
+            otherHalfTree = buildTree(
+                latentVertices,
+                probabilisticVertices,
+                logProbGradientCalculator,
+                sampleFromVertices,
+                currentTree.leapfrogBackward,
+                logU,
+                buildDirection,
+                treeHeight,
+                epsilon,
+                logOfMasterPMinusMomentumBeforeLeapfrog,
+                random
+            );
+
+            currentTree.leapfrogBackward = otherHalfTree.leapfrogBackward;
+
+        } else {
+
+            otherHalfTree = buildTree(
+                latentVertices,
+                probabilisticVertices,
+                logProbGradientCalculator,
+                sampleFromVertices,
+                currentTree.leapfrogForward,
+                logU,
+                buildDirection,
+                treeHeight,
+                epsilon,
+                logOfMasterPMinusMomentumBeforeLeapfrog,
+                random
+            );
+
+            currentTree.leapfrogForward = otherHalfTree.leapfrogForward;
+        }
+
+        return otherHalfTree;
+    }
+
+    public static TreeBuilder buildTree(List<Vertex<DoubleTensor>> latentVertices,
+                                        List<Vertex> probabilisticVertices,
+                                        LogProbGradientCalculator logProbGradientCalculator,
+                                        final List<? extends Vertex> sampleFromVertices,
+                                        Leapfrog leapfrog,
+                                        double logU,
+                                        int buildDirection,
+                                        int treeHeight,
+                                        double epsilon,
+                                        double logOfMasterPMinusMomentumBeforeLeapfrog,
+                                        KeanuRandom random) {
```

Comment:
you're right - but I'd rather keep these as parameters so that the function matches the pseudocode of the paper, or else it becomes very difficult (especially with the recursive function in there) to piece together what's going on

Reply:
ok